### DISCUSSION TITLE
Streamlining Repeated Worldgen?

### DISCUSSION THREAD

**Discussion by tmqCypher** [DFHack/dfhack#3774] [Ideas]
Hey folks, big fan and aspiring contributor here looking for some advice on a pain point of mine.

Being a benevolent god-manager, I like to hand pick a beautiful world for my dwarves, but DF's current UI makes repeated world generation tedious. When I get a world unbefitting my grand designs, I have to:
- pause history generation (or be forced to save the world once generation stops)
- back to main menu
- create new world
- re-select advanced parameters
- wait for DF to load mods/data

...before I can get another world going. It would be nice to skip the above and jump back to the start of map generation, especially with smaller worlds, which are typically more difficult to find desirable embark sites in.

I've checked the API/df-structures/forums and haven't found evidence of a clear way to hook into specific parts of the worldgen process (my thought being that I could pause the map's generation, evaluate it, and cause DF to reject it by borking its topology before it performs its final checks). I'd be willing to get into the weeds with reverse engineering if that seems necessary. 

I know this could be a lot of effort for incrementally more efficient micromanagement of a process that we already have a lot of control over, so I'm open to the suggestion that it's not worth the time, but I thought I'd float the idea anyways. If anyone knows what direction I should set off in, I'd greatly appreciate some pointers!

**Comment by myk002** [comment]
We have very little tooling around worldgen. Your approach sounds like a good one - generate, pause, inspect, reject, intentionally bork, repeat. I do not know how to do this safely myself, but others on the Discord server might know more.

How does PerfectWorld do it? I believe DF even has some commandline parameters to enable PerfectWorld's approach. Is there something there  you can piggyback on?

By the way, it probably wouldn't be too hard to auto-apply saved advanced parameter settings. check out `gui/mod-manager` for how it does it for mods.

**Reply by tmqCypher** [reply]
I hadn't found PerfectWorld before I posted, but it actually achieves the bulk of what I was hoping for, so thanks for bringing it to my attention! It just modifies your world_gen.txt file, which is simple enough that I could probably emulate some of its features in a script for the advanced parameters screen. Having more control from the outset is better than just being able to retry faster anyways.

As far as I can tell, the command line worldgen feature isn't working in v50+. Figuring out how it's supposed to work could be a good place to start if I want to go further with understanding the guts though.
